docs/machine-learning/basic-concepts-model-training-in-mldotnet.MD

```diff
@@ -0,0 +1,169 @@
+  - Each column has a *Name*, a *Type*, and an arbitrary set of *metadata* associated with it.
+  - It is important to note that one of the types is the `vector<T, N>` type, which means that the column's values are *vectors of items of type T, with the size of N*. This is a recommended way to represent multi-dimensional data associated with every row, like pixels in an image, or tokens in a text.
+  - The column's *metadata* contains information like 'slot names' of a vector column and suchlike. The metadata itself is actually represented as another one-row *data*, that is unique to each column.
+- The data view is a source of *cursors*. Think SQL cursors: a cursor is an object that iterates through the data, one row at a time, and presents the available data.
+  - Naturally, data can have as many active cursors over it as needed: since data itself is immutable, cursors are truly independent.
+  - Note that cursors typically access only a subset of columns: for efficiency, we do not compute the values of columns that are not 'needed' by the cursor.
+
+## Transformer
+
+A transformer is a component that takes data, does some work on it, and return new 'transformed' data.
+
+Here's the interface of `ITransformer`:
+
+```csharp
+public interface ITransformer
+{
+    IDataView Transform(IDataView input);
+    ISchema GetOutputSchema(ISchema inputSchema);
+}
+```
+
+As you can see, the transformer can `Transform` an input data to produce the output data. The other method, `GetOutputSchema`, is a mechanism of *schema propagation*: it allows you to see how the output data will look like for a given shape of the input data without actually performing the transformation.
+
+Most transformers in ML.NET tend to operate on one *input column* at a time, and produce the *output column*. For example a `new HashTransformer("foo", "bar")` would take the values from column "foo", hash them and put them into column "bar". 
+
+It is also common that the input and output column names are the same. In this case, the old column is 'replaced' with the new one. For example, a `new HashTransformer("foo")` would take the values from column "foo", hash them and 'put them back' into "foo". 
+
+Any transformer will, of course, produce a new data view when `Transform` is called: remember, data views are immutable.
+
+Another important consideration is that, because data is lazily evaluated, *transformers are lazy too*. Essentially, after you call
+
+```csharp
+var newData = transformer.Transform(oldData)
+```
+
+no actual computation will happen: only after you get a cursor from `newData` and start consuming the value will `newData` invoke the `transformer`'s transformation logic (and even that only if `transformer` in question is actually needed to produce the requested columns).
+
+### Transformer chains
+
+A useful property of a transformer is that *you can phrase a sequential application of transformers as yet another transformer*:
```

Comment:
It would be simpler to say that transformers support fluent composition, if the audience is expected to understand that term.